cli/compose/schema/data/config_schema_v3.1.json

```diff
@@ -167,7 +167,19 @@
         "ports": {
           "type": "array",
           "items": {
-            "type": ["string", "number"],
+            "oneOf": [
+              {"type": "number"},
+              {"type": "string"},
+              {
+                "type": "object",
+                "properties": {
+                  "mode": {"type": "string"},
+                  "target": {"type": "integer"},
+                  "published": {"type": "integer"},
+                  "protocol": {"type": "string"}
+                }
+              }
+            ],
             "format": "ports"
```

Comment:
I think you might need to move `format: ports` into the `type: number`, and `type: string` cases